### ISSUE TITLE
LTXFlowEuler: first denoising steps silently discarded due to bf16 comparison in tokens_to_denoise_mask

### ISSUE THREAD

**Issue by YiRen19** [NVlabs/Sana#423]
# SANA `LTXFlowEuler` BF16 Mask Bug

> 中文摘要：SANA 官方代码 `diffusion/scheduler/flow_euler_sampler.py` 的 `LTXFlowEuler.sample` 中，`tokens_to_denoise_mask` 的比较在 BF16 下执行，导致 t > 998.05 的去噪步（shift=9.8 的 60 步网格下为前 2 步）被静默丢弃——模型前向照算，但更新不会写回 latent。作者本意（由 `-1e-6` 可见）是让 t=1000 正常去噪。已在 RTX PRO 6000 + torch 2.12 + bf16 全程验证，本地 clone 为 commit `6298508`（该文件除我们自加的 callback 外与官方一致）。
>
> 影响：每次推理浪费约 2/60 次 DiT 前向（≈3.3% 算力）；有效去噪步数实为 58。
>
> 修复建议：将比较提升到 f32，例如 `tokens_to_denoise_mask = t / 1000 - 1e-6 < (1.0 - condition_mask.float())`。
>
> 注：上游最新 main 已核实（2026-07-17），bug 仍然存在，行号 131 与 187 与本地 commit `6298508` 一致。

---

## Bug Report (English, ready for GitHub issue)

**Title**: `LTXFlowEuler`: first denoising steps silently discarded due to bf16 comparison in `tokens_to_denoise_mask`

**Affected file**: `diffusion/scheduler/flow_euler_sampler.py`, class `LTXFlowEuler`, method `sample` (verified at local commit `6298508` and current upstream main).

### Summary

In `LTXFlowEuler.sample`, the per-step update is gated by

```python
tokens_to_denoise_mask = t / 1000 - 1e-6 < (1.0 - condition_mask)
latents = torch.where(tokens_to_denoise_mask, denoised_latents, latents)
```

`condition_mask` is created via `torch.zeros_like(latents)` and is therefore **bf16** under standard bf16 mixed-precision inference. `t / 1000 - 1e-6` is a **0-dim fp32 tensor**. PyTorch's type promotion demotes 0-dim tensors to the nd-tensor's dtype, so the comparison executes in bf16. bf16 has ~2^-8 resolution near 1.0, so any `t/1000 - 1e-6 >= 0.998046875` rounds up to exactly `1.0`, making `1.0 < 1.0` evaluate to **False for all tokens** — including generated (non-condition) tokens. The scheduler's update for those steps is computed but entirely discarded.

### Empirical evidence (SANA-WM bidirectional, 161 frames, 60 steps, cfg 5.0, flow_shift 9.8, bf16)

1. The saved latent trajectory satisfies `z_0 == z_1 == z_init` **bitwise** — the first two loop iterations leave the latent completely unchanged.
2. Quantizing a linear layer changes `noise_pred` at step 0 substantially (MSE ≈ 1.8), yet the latent after step 0 remains bit-identical to baseline.
3. Instrumenting `FlowMatchEulerDiscreteScheduler.step` shows `dt = 0.00174 ≠ 0` at step 0 — the scheduler produces a nonzero update that never reaches the latent.
4. Reproducing the mask expression on CPU with the exact dtype chain gives:
   - t = 1000.0 → `t/1000 - 1e-6 = 0.99999899` (f32) → cast to bf16 → `1.0` → mask **False**
   - t = 998.26 → bf16(`0.99825996`) = `1.0` → mask **False**
   - t = 996.47 → bf16(`0.99646699`) = `0.99609375` → `< 1.0` → mask **True**

   This exactly matches the observed "frozen steps 0–1, active from step 2".
5. The `-1e-6` term indicates the author intended `t/1000 ≤ 1` to imply denoise (i.e. step 0 was meant to be executed); in f32 the expression behaves as intended.

### Impact

- With the SANA-WM config (60 steps, `inference_flow_shift=9.8`): steps **0 and 1** are no-ops → ~3.3% of DiT compute wasted per inference; effective denoising is 58 steps.
- Affected-step count depends on the timestep grid. With shift=8.0 or 3.0, only step 0 is dropped. In general any step with `t > 998.046875` is dropped (bf16 rounding boundary below 1.0 is `0.99609375`, midpoint `0.998046875`).
- Quality impact is presumably minor (the pipeline has always behaved this way), but sampler semantics silently differ from the intended schedule, and compute is wasted.

### Suggested fix

Compute the comparison in f32, e.g. either of:

```python
tokens_to_denoise_mask = t / 1000 - 1e-6 < (1.0 - condition_mask.float())
# or
condition_mask = torch.zeros(latents.shape, dtype=torch.float32, device=latents.device)
```

### Minimal repro (CPU)

```python
import torch

condition_mask = torch.zeros(1, 4, 2, 3, 3, dtype=torch.bfloat16)  # zeros_like(latents)
for tval in [1000.0, 998.26, 996.47]:
    t = torch.tensor(tval, dtype=torch.float32)                     # 0-dim, like enumerate(timesteps)
    mask = t / 1000 - 1e-6 < (1.0 - condition_mask)                 # demoted to bf16
    print(tval, "-> generated-token mask:", bool(mask[0, 0, 1, 0, 0]))

# 1000.0 -> False
# 998.26 -> False
# 996.47 -> True
# Expected: True in all three cases, since -1e-6 intends t <= 1000 to denoise.
```
